File: android/app/src/main/java/com/ayuvo/health/records/model/RecordIntelligence.kt
package com.ayuvo.health.records.model

/**
 * Phase 2 "Intelligence" value types (docs/health-records.md §8–§17). Every enum stores the exact
 * lowercase contract string; unknown stored values fall back to a safe default.
 */

/** `record_fields.field_key` (§8). */
object FieldKey {
    const val DOCTOR_NAME = "doctor_name"
    const val DOCTOR_SPECIALTY = "doctor_specialty"
    const val FACILITY = "facility"
    const val DEPARTMENT = "department"
    const val PATIENT_NAME = "patient_name"
    const val PATIENT_AGE = "patient_age"
    const val PATIENT_SEX = "patient_sex"
    const val REPORT_NAME = "report_name"
    const val TEST_RESULT = "test_result"
    const val DIAGNOSIS = "diagnosis"
    const val SYMPTOM = "symptom"
    const val MEDICATION = "medication"
    const val PROCEDURE = "procedure"
    const val RECOMMENDATION = "recommendation"
    const val FOLLOW_UP_DATE = "follow_up_date"
    const val VISIT_DATE = "visit_date"
    const val COLLECTION_DATE = "collection_date"
    const val REPORT_DATE = "report_date"
    const val PRESCRIPTION_DATE = "prescription_date"
    const val ADMISSION_DATE = "admission_date"
    const val DISCHARGE_DATE = "discharge_date"
    const val DOCUMENT_TIME = "document_time"
    const val LOCATION = "location"

    val ALL: List<String> = listOf(
        DOCTOR_NAME, DOCTOR_SPECIALTY, FACILITY, DEPARTMENT, PATIENT_NAME, PATIENT_AGE, PATIENT_SEX,
        REPORT_NAME, TEST_RESULT, DIAGNOSIS, SYMPTOM, MEDICATION, PROCEDURE, RECOMMENDATION,
        FOLLOW_UP_DATE, VISIT_DATE, COLLECTION_DATE, REPORT_DATE, PRESCRIPTION_DATE, ADMISSION_DATE,
        DISCHARGE_DATE, DOCUMENT_TIME, LOCATION
    )

    /** Keys that legitimately hold several values; every other key is single-valued (§8.1). */
    val MULTI_VALUED: Set<String> = setOf(TEST_RESULT, DIAGNOSIS, SYMPTOM, MEDICATION, PROCEDURE, RECOMMENDATION)

    val DATE_KEYS: Set<String> = setOf(
        FOLLOW_UP_DATE, VISIT_DATE, COLLECTION_DATE, REPORT_DATE, PRESCRIPTION_DATE, ADMISSION_DATE, DISCHARGE_DATE
    )

    /** Order used for `document_date` derivation (§8.1). */
    val DOCUMENT_DATE_ORDER: List<String> = listOf(REPORT_DATE, COLLECTION_DATE, PRESCRIPTION_DATE, DISCHARGE_DATE, VISIT_DATE)

    /** Key fields reviewed when uncertain (§15). */
    fun isKeyField(key: String): Boolean = key == REPORT_NAME || key == DOCTOR_NAME || key == FACILITY || key in DATE_KEYS

    fun isSingleValued(key: String): Boolean = key !in MULTI_VALUED
}

enum class ExtractionMethod(val raw: String) {
    FILE_METADATA("file_metadata"), PDF_TEXT("pdf_text"), OCR("ocr"), RULES("rules"),
    AI_LOCAL("ai_local"), AI_CLOUD("ai_cloud"), USER("user");

    val isAi: Boolean get() = this == AI_LOCAL || this == AI_CLOUD

    companion object {
        fun fromRaw(raw: String?): ExtractionMethod = entries.firstOrNull { it.raw == raw } ?: RULES
    }
}

enum class FieldState(val raw: String) {
    SUGGESTED("suggested"), CONFIRMED("confirmed"), REJECTED("rejected"), USER("user");

    /** Rows the pipeline never touches again (§8.1). */
    val isSettled: Boolean get() = this != SUGGESTED

    companion object {
        fun fromRaw(raw: String?): FieldState = entries.firstOrNull { it.raw == raw } ?: SUGGESTED
    }
}

enum class ResultFlag(val raw: String) {
    LOW("low"), HIGH("high"), CRITICAL_LOW("critical_low"), CRITICAL_HIGH("critical_high"),
    NORMAL("normal"), ABNORMAL("abnormal"), UNKNOWN("unknown");

    /** Anything the report marks outside its range. */
    val isAbnormal: Boolean get() = this != NORMAL && this != UNKNOWN

    companion object {
        fun fromRaw(raw: String?): ResultFlag = entries.firstOrNull { it.raw == raw } ?: UNKNOWN
    }
}

enum class HighlightSection(val raw: String) {
    IMPORTANT("important"), MEDICATIONS("medications"), RECOMMENDATIONS("recommendations"), SUMMARY("summary");

    companion object {
        fun fromRaw(raw: String?): HighlightSection = entries.firstOrNull { it.raw == raw } ?: IMPORTANT
    }
}

/** `processing_jobs.stage`, in pipeline order (§8, §9; Phase 3 adds `observations` and `relations`). */
enum class ProcessingStage(val raw: String) {
    TEXT("text"), CLASSIFY("classify"), BOUNDARIES("boundaries"), RULES("rules"), AI("ai"),
    VALIDATE("validate"), HIGHLIGHTS("highlights"), REVIEW("review"), NEAR_DUPLICATE("near_duplicate"),
    OBSERVATIONS("observations"), RELATIONS("relations"), INDEX("index"), DONE("done");

    fun next(): ProcessingStage = entries.getOrElse(ordinal + 1) { DONE }

    companion object {
        fun fromRaw(raw: String?): ProcessingStage = entries.firstOrNull { it.raw == raw } ?: TEXT
    }
}

/** `records.ai_mode_used`. */
enum class AiModeUsed(val raw: String) {
    NONE("none"), LOCAL("local"), CLOUD("cloud");

    companion object {
        fun fromRaw(raw: String?): AiModeUsed = entries.firstOrNull { it.raw == raw } ?: NONE
    }
}

/** Preference `healthRecordsAiMode` (§16). Null (unset) is handled by callers. */
enum class RecordsAiMode(val raw: String) {
    LOCAL("local"), CLOUD("cloud"), ASK("ask"), OFF("off");

    companion object {
        fun fromRaw(raw: String?): RecordsAiMode? = entries.firstOrNull { it.raw == raw }
    }
}

enum class DuplicateReason(val raw: String) {
    CHECKSUM("checksum"), PHASH("phash"), CONTENT("content");

    companion object {
        fun fromRaw(raw: String?): DuplicateReason = entries.firstOrNull { it.raw == raw } ?: CONTENT
    }
}

enum class DuplicateResolution(val raw: String) {
    PENDING("pending"), KEEP_BOTH("keep_both"), REPLACED("replaced"), MERGED("merged"), CANCELLED("cancelled");

    companion object {
        fun fromRaw(raw: String?): DuplicateResolution = entries.firstOrNull { it.raw == raw } ?: PENDING
    }
}

enum class SplitStatus(val raw: String) {
    PENDING("pending"), ACCEPTED("accepted"), REJECTED("rejected");

    companion object {
        fun fromRaw(raw: String?): SplitStatus = entries.firstOrNull { it.raw == raw } ?: PENDING
    }
}

/** `processing_error` values (§9). */
object ProcessingError {
    const val TEXT_UNAVAILABLE = "text_unavailable"
    const val OCR_FAILED = "ocr_failed"
    const val AI_FAILED = "ai_failed"
    const val AI_UNAVAILABLE = "ai_unavailable"
    const val PROTECTED_PDF = "protected_pdf"
    const val UNSUPPORTED = "unsupported"
}

/** One `record_fields` row. */
data class RecordField(
    val id: String,
    val recordId: String,
    val key: String,
    val valueText: String,
    val valueJson: String? = null,
    val method: ExtractionMethod,
    val confidence: Double,
    val state: FieldState = FieldState.SUGGESTED,
    val sourcePage: Int? = null,
    val sourceBbox: String? = null,
    val evidence: String? = null,
    val createdMs: Long = 0,
    val updatedMs: Long = 0
)

/** A value produced by rules or AI before [com.ayuvo.health.records.processing.ExtractionWriter] decides what to store. */
data class ExtractedField(
    val key: String,
    val valueText: String,
    val valueJson: String? = null,
    val method: ExtractionMethod,
    val confidence: Double,
    val sourcePage: Int? = null,
    val sourceBbox: String? = null,
    val evidence: String? = null
)

/** One `record_highlights` row. */
data class RecordHighlight(
    val id: String,
    val recordId: String,
    val section: HighlightSection,
    val text: String,
    val method: ExtractionMethod,
    val provider: String? = null,
    val fieldId: String? = null,
    val sourcePage: Int? = null,
    val confidence: Double = 0.0,
    val dismissed: Boolean = false,
    val position: Int = 0,
    val createdMs: Long = 0
)

/** One `processing_jobs` row. */
data class ProcessingJob(
    val recordId: String,
    val stage: ProcessingStage,
    val attempts: Int = 0,
    val nextAttemptMs: Long = 0,
    val lastError: String? = null,
    /** Per-record AI choice from the Ask flow (`local` / `cloud` / `off`), else null = preference. */
    val requestedMode: String? = null,
    val awaitingConsent: Boolean = false,
    val updatedMs: Long = 0
)

data class DuplicateCandidate(
    val recordId: String,
    val existingId: String,
    val reason: DuplicateReason,
    val score: Double,
    val resolution: DuplicateResolution = DuplicateResolution.PENDING,
    val createdMs: Long = 0
)

/** `split_proposals.segments_json` entry; pages are 0-based inclusive. */
data class SplitSegment(
    val pageStart: Int,
    val pageEnd: Int,
    val recordType: RecordType,
    val title: String,
    val confidence: Double
)

data class SplitProposal(
    val recordId: String,
    val segments: List<SplitSegment>,
    val status: SplitStatus = SplitStatus.PENDING,
    val createdMs: Long = 0,
    val updatedMs: Long = 0
)

/** Intelligence columns of `records` (v2) that the list projection does not carry. */
data class RecordIntelligenceColumns(
    val phash: String? = null,
    val textSignature: String? = null,
    val aiModeUsed: AiModeUsed = AiModeUsed.NONE,
    val aiProvider: String? = null,
    val typeConfidence: Double? = null,
    val typeMethod: ExtractionMethod? = null
)

File: ios/calorietracker/Records/Models/RecordIntelligence.swift
import Foundation

// Phase 2 "Intelligence" enumerations and rows (docs/health-records.md §8, §9, §15, §16).
// Every raw value is the exact lowercase string stored in SQLite.

nonisolated enum RecordFieldKey: String, CaseIterable, Codable, Sendable, Identifiable {
    case doctorName = "doctor_name"
    case doctorSpecialty = "doctor_specialty"
    case facility
    case department
    case patientName = "patient_name"
    case patientAge = "patient_age"
    case patientSex = "patient_sex"
    case reportName = "report_name"
    case testResult = "test_result"
    case diagnosis
    case symptom
    case medication
    case procedure
    case recommendation
    case followUpDate = "follow_up_date"
    case visitDate = "visit_date"
    case collectionDate = "collection_date"
    case reportDate = "report_date"
    case prescriptionDate = "prescription_date"
    case admissionDate = "admission_date"
    case dischargeDate = "discharge_date"
    case documentTime = "document_time"
    case location

    var id: String { rawValue }

    /// Keys that may legitimately hold several values (§8.1); every other key is single-valued.
    static let multiValued: Set<RecordFieldKey> = [.testResult, .diagnosis, .symptom, .medication, .procedure, .recommendation]

    static let dateKeys: Set<RecordFieldKey> = [.followUpDate, .visitDate, .collectionDate, .reportDate, .prescriptionDate, .admissionDate, .dischargeDate]

    /// §8.1 document-date priority.
    static let documentDatePriority: [RecordFieldKey] = [.reportDate, .collectionDate, .prescriptionDate, .dischargeDate, .visitDate]

    /// §15 key fields for review.
    static let reviewKeyFields: Set<RecordFieldKey> = Set<RecordFieldKey>([.reportName, .doctorName, .facility]).union(dateKeys)

    var isMultiValued: Bool { Self.multiValued.contains(self) }
    var isDate: Bool { Self.dateKeys.contains(self) }

    var title: String {
        switch self {
        case .doctorName: String(localized: "Doctor")
        case .doctorSpecialty: String(localized: "Specialty")
        case .facility: String(localized: "Hospital or lab")
        case .department: String(localized: "Department")
        case .patientName: String(localized: "Patient")
        case .patientAge: String(localized: "Age")
        case .patientSex: String(localized: "Sex")
        case .reportName: String(localized: "Report")
        case .testResult: String(localized: "Test result")
        case .diagnosis: String(localized: "Diagnosis")
        case .symptom: String(localized: "Symptom")
        case .medication: String(localized: "Medication")
        case .procedure: String(localized: "Procedure")
        case .recommendation: String(localized: "Recommendation")
        case .followUpDate: String(localized: "Follow-up date")
        case .visitDate: String(localized: "Visit date")
        case .collectionDate: String(localized: "Collection date")
        case .reportDate: String(localized: "Report date")
        case .prescriptionDate: String(localized: "Prescription date")
        case .admissionDate: String(localized: "Admission date")
        case .dischargeDate: String(localized: "Discharge date")
        case .documentTime: String(localized: "Time")
        case .location: String(localized: "Location")
        }
    }

    /// Detail-screen grouping of "Extracted information".
    var group: RecordFieldGroup {
        switch self {
        case .doctorName, .doctorSpecialty, .facility, .department: .care
        case .patientName, .patientAge, .patientSex: .patient
        case .reportName, .diagnosis, .symptom, .procedure, .location: .clinical
        case .testResult: .results
        case .medication: .medications
        case .recommendation: .recommendations
        case .followUpDate, .visitDate, .collectionDate, .reportDate, .prescriptionDate, .admissionDate, .dischargeDate, .documentTime: .dates
        }
    }
}

nonisolated enum RecordFieldGroup: String, CaseIterable, Sendable, Identifiable {
    case dates, care, patient, clinical, results, medications, recommendations

    var id: String { rawValue }

    var title: String {
        switch self {
        case .dates: String(localized: "Dates")
        case .care: String(localized: "Doctor & facility")
        case .patient: String(localized: "Patient")
        case .clinical: String(localized: "Clinical details")
        case .results: String(localized: "Test results")
        case .medications: String(localized: "Medications")
        case .recommendations: String(localized: "Recommendations")
        }
    }
}

/// `method` of fields, highlights and `records.type_method` (§8).
nonisolated enum RecordFieldMethod: String, CaseIterable, Codable, Sendable {
    case fileMetadata = "file_metadata"
    case pdfText = "pdf_text"
    case ocr
    case rules
    case aiLocal = "ai_local"
    case aiCloud = "ai_cloud"
    case user

    var isAI: Bool { self == .aiLocal || self == .aiCloud }

    var badgeTitle: String {
        switch self {
        case .fileMetadata: String(localized: "File details")
        case .pdfText, .ocr, .rules: String(localized: "Rules")
        case .aiLocal: String(localized: "On-device AI")
        case .aiCloud: String(localized: "Online AI")
        case .user: String(localized: "You")
        }
    }

    var dateMethod: RecordDateMethod {
        switch self {
        case .fileMetadata: .fileMetadata
        case .pdfText: .pdfText
        case .ocr: .ocr
        case .rules: .rules
        case .aiLocal: .aiLocal
        case .aiCloud: .aiCloud
        case .user: .user
        }
    }
}

nonisolated enum RecordFieldState: String, CaseIterable, Codable, Sendable {
    case suggested
    case confirmed
    case rejected
    case user

    /// Rows a new extraction never touches (§8.1 step 1).
    var isLocked: Bool { self != .suggested }
}

nonisolated enum RecordResultFlag: String, CaseIterable, Codable, Sendable {
    case low
    case high
    case criticalLow = "critical_low"
    case criticalHigh = "critical_high"
    case normal
    case abnormal
    case unknown

    var isAbnormal: Bool {
        switch self {
        case .low, .high, .criticalLow, .criticalHigh, .abnormal: true
        case .normal, .unknown: false
        }
    }

    var symbol: String? {
        switch self {
        case .low, .criticalLow: "↓"
        case .high, .criticalHigh: "↑"
        case .abnormal: "*"
        case .normal, .unknown: nil
        }
    }
}

nonisolated enum RecordHighlightSection: String, CaseIterable, Codable, Sendable {
    case important
    case medications
    case recommendations
    case summary
}

nonisolated enum RecordAIModeUsed: String, CaseIterable, Codable, Sendable {
    case none
    case local
    case cloud
}

nonisolated enum RecordProcessingStage: String, CaseIterable, Codable, Sendable {
    case text
    case classify
    case boundaries
    case rules
    case ai
    case validate
    case highlights
    case review
    case nearDuplicate = "near_duplicate"
    /// Phase 3: promotion + analyte mapping (§19).
    case observations
    /// Phase 3: relation suggestions (§22).
    case relations
    case index
    case done

    var next: RecordProcessingStage {
        let all = Self.allCases
        guard let index = all.firstIndex(of: self), index + 1 < all.count else { return .done }
        return all[index + 1]
    }

    /// `processing_status` shown while this stage runs (§9).
    var runningStatus: RecordProcessingStatus {
        switch self {
        case .text: .extractingText
        case .done: .ready
        default: .analyzing
        }
    }
}

nonisolated enum RecordDuplicateReason: String, CaseIterable, Codable, Sendable {
    case checksum
    case phash
    case content

    var title: String {
        switch self {
        case .checksum: String(localized: "Identical file")
        case .phash: String(localized: "Looks the same")
        case .content: String(localized: "Same content")
        }
    }
}

nonisolated enum RecordDuplicateResolution: String, CaseIterable, Codable, Sendable {
    case pending
    case keepBoth = "keep_both"
    case replaced
    case merged
    case cancelled
}

nonisolated enum RecordSplitStatus: String, CaseIterable, Codable, Sendable {
    case pending
    case accepted
    case rejected
}

/// `processing_error` values (§9).
nonisolated enum RecordProcessingError: String, CaseIterable, Sendable {
    case textUnavailable = "text_unavailable"
    case ocrFailed = "ocr_failed"
    case aiFailed = "ai_failed"
    case aiUnavailable = "ai_unavailable"
    case protectedPDF = "protected_pdf"
    case unsupported
}

/// `healthRecordsAiMode` (§6b, §16). Unset is represented by `nil`.
nonisolated enum RecordsAIMode: String, CaseIterable, Codable, Sendable, Identifiable {
    case local
    case cloud
    case ask
    case off

    static let storageKey = "healthRecordsAiMode"

    var id: String { rawValue }

    var title: String {
        switch self {
        case .local: String(localized: "Local AI")
        case .cloud: String(localized: "Cloud AI")
        case .ask: String(localized: "Ask me when needed")
        case .off: String(localized: "Skip AI for now")
        }
    }

    var subtitle: String {
        switch self {
        case .local: String(localized: "Uses the AI on this iPhone. Nothing is sent online.")
        case .cloud: String(localized: "Sends page text to your AI provider when details are missing.")
        case .ask: String(localized: "Ayuvo asks before using AI on a record.")
        case .off: String(localized: "Details are found with on-device rules only.")
        }
    }

    var systemImage: String {
        switch self {
        case .local: "iphone"
        case .cloud: "cloud"
        case .ask: "questionmark.bubble"
        case .off: "nosign"
        }
    }

    static func stored(in defaults: UserDefaults = .standard) -> RecordsAIMode? {
        defaults.string(forKey: storageKey).flatMap(RecordsAIMode.init(rawValue:))
    }
}

// MARK: - Rows

/// One `record_fields` row.
nonisolated struct RecordField: Identifiable, Hashable, Sendable {
    var id: String
    var recordID: String
    var key: RecordFieldKey
    var valueText: String
    var valueJSON: String?
    var method: RecordFieldMethod
    var confidence: Double
    var state: RecordFieldState
    var sourcePage: Int?
    var sourceBBox: String?
    var evidence: String?
    var createdMs: Int64
    var updatedMs: Int64

    var testResult: RecordTestResultValue? {
        guard key == .testResult, let valueJSON else { return nil }
        return RecordTestResultValue.decode(valueJSON)
    }

    var medicationValue: RecordMedicationValue? {
        guard key == .medication, let valueJSON else { return nil }
        return RecordMedicationValue.decode(valueJSON)
    }

    /// `[x, y, w, h]` normalized top-left, when stored.
    var bbox: [Double]? {
        guard let sourceBBox, let data = sourceBBox.data(using: .utf8),
              let values = try? JSONSerialization.jsonObject(with: data) as? [NSNumber], values.count == 4
        else { return nil }
        return values.map(\.doubleValue)
    }

    var isRoleReferrer: Bool {
        guard key == .doctorName, let valueJSON, let data = valueJSON.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return false }
        return (object["role"] as? String) == "referrer"
    }

    /// Human text for rows and review.
    var displayValue: String {
        if let result = testResult {
            var parts = [result.value ?? valueText]
            if let unit = result.unit, !unit.isEmpty { parts.append(unit) }
            if let symbol = result.flag?.symbol { parts.append(symbol) }
            return parts.joined(separator: " ")
        }
        if let medication = medicationValue {
            return [medication.name ?? valueText, medication.strength, medication.frequency, medication.duration]
                .compactMap { $0 }
                .filter { !$0.isEmpty }
                .joined(separator: " · ")
        }
        if key.isDate, let date = RecordDates.date(fromDay: valueText) {
            return date.formatted(date: .abbreviated, time: .omitted)
        }
        return valueText
    }
}

nonisolated struct RecordTestResultValue: Codable, Hashable, Sendable {
    var name: String?
    var value: String?
    var valueNum: Double?
    var unit: String?
    var refText: String?
    var refLow: Double?
    var refHigh: Double?
    var flag: RecordResultFlag?

    enum CodingKeys: String, CodingKey {
        case name, value, unit, flag
        case valueNum = "value_num"
        case refText = "ref_text"
        case refLow = "ref_low"
        case refHigh = "ref_high"
    }

    static func decode(_ json: String) -> RecordTestResultValue? {
        guard let data = json.data(using: .utf8) else { return nil }
        return try? JSONDecoder().decode(RecordTestResultValue.self, from: data)
    }
}

nonisolated struct RecordMedicationValue: Codable, Hashable, Sendable {
    var name: String?
    var strength: String?
    var form: String?
    var dose: String?
    var frequency: String?
    var duration: String?
    var instructions: String?

    static func decode(_ json: String) -> RecordMedicationValue? {
        guard let data = json.data(using: .utf8) else { return nil }
        return try? JSONDecoder().decode(RecordMedicationValue.self, from: data)
    }
}

/// One `record_highlights` row.
nonisolated struct RecordHighlight: Identifiable, Hashable, Sendable {
    var id: String
    var recordID: String
    var section: RecordHighlightSection
    var text: String
    var method: RecordFieldMethod
    var provider: String?
    var fieldID: String?
    var sourcePage: Int?
    var confidence: Double
    var dismissed: Bool
    var position: Int
    var createdMs: Int64
}

/// One `processing_jobs` row.
nonisolated struct RecordProcessingJob: Hashable, Sendable {
    var recordID: String
    var stage: RecordProcessingStage
    var attempts: Int
    var nextAttemptMs: Int64
    var lastError: String?
    var requestedMode: String?
    var awaitingConsent: Bool
    var updatedMs: Int64
}

/// One `duplicate_candidates` row.
nonisolated struct RecordDuplicateCandidate: Hashable, Sendable, Identifiable {
    var recordID: String
    var existingID: String
    var reason: RecordDuplicateReason
    var score: Double
    var resolution: RecordDuplicateResolution
    var createdMs: Int64

    var id: String { recordID + "|" + existingID }
}

nonisolated struct RecordSplitSegment: Codable, Hashable, Sendable, Identifiable {
    var pageStart: Int
    var pageEnd: Int
    var recordType: RecordType
    var title: String
    var confidence: Double

    var id: String { "\(pageStart)-\(pageEnd)" }

    enum CodingKeys: String, CodingKey {
        case pageStart = "page_start"
        case pageEnd = "page_end"
        case recordType = "record_type"
        case title
        case confidence
    }

    init(pageStart: Int, pageEnd: Int, recordType: RecordType, title: String, confidence: Double) {
        self.pageStart = pageStart
        self.pageEnd = pageEnd
        self.recordType = recordType
        self.title = title
        self.confidence = confidence
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        pageStart = try container.decode(Int.self, forKey: .pageStart)
        pageEnd = try container.decode(Int.self, forKey: .pageEnd)
        recordType = RecordType(rawValue: (try? container.decode(String.self, forKey: .recordType)) ?? "") ?? .other
        title = (try? container.decode(String.self, forKey: .title)) ?? ""
        confidence = (try? container.decode(Double.self, forKey: .confidence)) ?? 0
    }
}

/// One `split_proposals` row.
nonisolated struct RecordSplitProposal: Hashable, Sendable {
    var recordID: String
    var segments: [RecordSplitSegment]
    var status: RecordSplitStatus
    var createdMs: Int64
    var updatedMs: Int64

    static func encodeSegments(_ segments: [RecordSplitSegment]) -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        return (try? encoder.encode(segments)).flatMap { String(data: $0, encoding: .utf8) } ?? "[]"
    }

    static func decodeSegments(_ json: String) -> [RecordSplitSegment] {
        guard let data = json.data(using: .utf8) else { return [] }
        return (try? JSONDecoder().decode([RecordSplitSegment].self, from: data)) ?? []
    }
}

// MARK: - Extraction input

/// One value found by rules or AI, before `applyExtraction` (§8.1).
nonisolated struct ExtractedField: Hashable, Sendable {
    var key: RecordFieldKey
    var valueText: String
    /// Compact JSON object text (sorted keys) or nil.
    var valueJSON: String?
    var method: RecordFieldMethod
    var confidence: Double
    var sourcePage: Int?
    var sourceBBox: [Double]?
    var evidence: String?

    init(
        key: RecordFieldKey,
        valueText: String,
        valueJSON: String? = nil,
        method: RecordFieldMethod = .rules,
        confidence: Double,
        sourcePage: Int? = nil,
        sourceBBox: [Double]? = nil,
        evidence: String? = nil
    ) {
        self.key = key
        self.valueText = valueText
        self.valueJSON = valueJSON
        self.method = method
        self.confidence = confidence
        self.sourcePage = sourcePage
        self.sourceBBox = sourceBBox
        self.evidence = evidence
    }
}

nonisolated struct ExtractedHighlight: Hashable, Sendable {
    var section: RecordHighlightSection
    var text: String
    var method: RecordFieldMethod
    var provider: String?
    var sourcePage: Int?
    var confidence: Double
}

nonisolated struct RecordExtraction: Sendable {
    var fields: [ExtractedField] = []
    /// Classifier / AI type suggestion, applied when `type_method` is not `user`.
    var recordType: RecordType?
    var typeConfidence: Double?
    var typeMethod: RecordFieldMethod?
    var summary: ExtractedHighlight?
}

/// Row counts for the processing strip.
nonisolated struct RecordsProcessingSummary: Hashable, Sendable {
    var activeRecords: Int = 0
    var awaitingConsent: Int = 0
    var needsReview: Int = 0
    var currentRecordID: String?
    var currentPage: Int?
    var currentPageCount: Int?
}
